domain/agentprovisioner/service/service_test.go

```diff
@@ -0,0 +1,218 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/errors"
+	"github.com/juju/proxy"
+	jc "github.com/juju/testing/checkers"
+	"go.uber.org/mock/gomock"
+	gc "gopkg.in/check.v1"
+
+	"github.com/juju/juju/core/containermanager"
+	"github.com/juju/juju/core/instance"
+	"github.com/juju/juju/core/model"
+	modeltesting "github.com/juju/juju/core/model/testing"
+	"github.com/juju/juju/environs/config"
+	"github.com/juju/juju/rpc/params"
+)
+
+type suite struct {
+	state          *MockState
+	provider       *MockProvider
+	providerGetter func(context.Context) (Provider, error)
+	modelID        model.UUID
+	service        *Service
+}
+
+var _ = gc.Suite(&suite{})
+
+func (s *suite) SetUpTest(c *gc.C) {
+	s.modelID = modeltesting.GenModelUUID(c)
+	// Default provider getter function
+	s.providerGetter = func(ctx context.Context) (Provider, error) {
+		return s.provider, nil
+	}
+}
+
+func (s *suite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+	s.state = NewMockState(ctrl)
+	s.provider = NewMockProvider(ctrl)
+	return ctrl
+}
+
+func (s *suite) setupService(c *gc.C) {
+	s.service = NewService(s.modelID, s.state, s.providerGetter)
```

Comment:
Is this needed? Just make the service in each test. Seems like an over optimisation.